src/core/task/tools/handlers/WriteToFileToolHandler.ts

```diff
@@ -187,223 +83,167 @@ export class WriteToFileToolHandler implements IFullyManagedTool {
-				if (text || (images && images.length > 0) || (files && files.length > 0)) {
-					let fileContentString = ""
-					if (files && files.length > 0) {
-						fileContentString = await processFilesIntoText(files)
-					}
-
-					// Push additional tool feedback using existing utilities
-					ToolResultUtils.pushAdditionalToolFeedback(
-						config.taskState.userMessageContent,
-						text,
-						images,
-						fileContentString,
-					)
-					await config.callbacks.say("user_feedback", text, images, files)
-				}
-
-				telemetryService.captureToolUsage(config.ulid, block.name, config.api.getModel().id, false, true)
-			}
-		}
-
 		try {
-			// Construct newContent from diff or content
-			let newContent: string = ""
-
-			if (diff) {
-				// Handle replace_in_file with diff construction
-				if (!config.api.getModel().id.includes("claude")) {
-					// deepseek models tend to use unescaped html entities in diffs
-					diff = fixModelHtmlEscaping(diff)
-					diff = removeInvalidChars(diff)
-				}
-
-				// Open the editor if not done already
-				if (!config.services.diffViewProvider.isEditing) {
-					await config.services.diffViewProvider.open(relPath)
-				}
-
-				try {
-					newContent = await constructNewFileContent(
-						diff,
-						config.services.diffViewProvider.originalContent || "",
-						true, // isFinal = true since we're not streaming
-					)
-				} catch (error) {
-					// Show diff error UI message (from original implementation)
-					await config.callbacks.say("diff_error", relPath)
-
-					// Extract error type from error message if possible
-					const errorType =
-						error instanceof Error && error.message.includes("does not match anything")
-							? "search_not_found"
-							: "other_diff_error"
-
-					// Add telemetry for diff edit failure
-					telemetryService.captureDiffEditFailure(config.ulid, config.api.getModel().id, errorType)
-
-					await config.services.diffViewProvider.revertChanges()
-					await config.services.diffViewProvider.reset()
-
-					// Return detailed error with original content for context
-					return formatResponse.toolError(
-						`${(error as Error)?.message}\n\n` +
-							formatResponse.diffError(relPath, config.services.diffViewProvider.originalContent),
-					)
-				}
-			} else if (content) {
-				// Handle write_to_file and new_rule with direct content
-				newContent = content
-
-				// Pre-processing newContent for cases where weaker models might add artifacts
-				if (newContent.startsWith("```")) {
-					newContent = newContent.split("\n").slice(1).join("\n").trim()
-				}
-				if (newContent.endsWith("```")) {
-					newContent = newContent.split("\n").slice(0, -1).join("\n").trim()
-				}
-
-				if (!config.api.getModel().id.includes("claude")) {
-					newContent = fixModelHtmlEscaping(newContent)
-					newContent = removeInvalidChars(newContent)
-				}
+			const result = await this.validateAndPrepareFileOperation(config, block, rawRelPath, rawDiff, rawContent)
```

Comment:
@saoudrizwan  Running this here might not be super necessary as the previous partial function covers the logic already but that being said this extra run is harmless as this run has already been done MANY times in the streaming logic.